src/Target.cpp

```diff
@@ -390,34 +333,105 @@ bool Target::merge_string(const std::string &target) {
     }
 
     // If arch is PNaCl, require explicit setting of os and bits as well.
-    if (arch_specified && arch == Target::PNaCl) {
-        if (!os_specified || os != Target::NaCl) {
+    if (arch_specified && t.arch == Target::PNaCl) {
+        if (!os_specified || t.os != Target::NaCl) {
             return false;
         }
-        if (!bits_specified || bits != 32) {
+        if (!bits_specified || t.bits != 32) {
             return false;
         }
     }
 
     return true;
 }
 
+void bad_target_string(const std::string &target) {
+    const char *separator = "";
+    std::string architectures;
+    for (const auto &arch_entry : arch_name_map) {
+        architectures += separator + arch_entry.first;
+        separator = ", ";
+    }
+    separator = "";
+    std::string oses;
+    for (auto os_entry : os_name_map) {
+        oses += separator + os_entry.first;
+        separator = ", ";
+    }
+    separator = "";
+    // Format the features to go one feature over 70 characters per line,
+    // assume the first line starts with "Features are ".
+    int line_char_start = -(int)sizeof("Features are");
+    std::string features;
+    for (auto feature_entry : feature_name_map) {
+        features += separator + feature_entry.first;
+        if (features.length() - line_char_start > 70) {
+            separator = "\n";
+            line_char_start = features.length();
+        } else {
+            separator = ", ";
+        }
+    }
+    user_error << "Did not understand HL_TARGET=" << target << "\n"
```

Comment:
nit: target strings can be specified in ways other than HL_TARGET (e.g., as arguments on the command line to a generator)